x-pack/plugin/ql/src/main/java/org/elasticsearch/xpack/ql/optimizer/OptimizerRules.java

```diff
@@ -1284,7 +1246,12 @@ Expression doSimplify() {
                     ? Literal.of(bcLiteral, ((Number) bcLiteral.value()).doubleValue())
                     : bcLiteral;
 
-                Expression bcRightExpression = safeMaybeFold(operation.inverse(bcl.source(), bcl, opRight));
+                if (operation instanceof BinaryComparisonInvertible == false) { // should be an assertion?
+                    throw new QlIllegalArgumentException("Unexpected [" + operation.symbol() + "] operation object");
+                }
```

Comment:
I've removed the check.

Offtopic:
> no need to double check the code flow

Extensive testing should definitely help. I believe assertions exist still because code has a (life) cycle and effects of later changes can't always be exhaustively checked/predicted or coded in (see the many `instanceof`s or [comment](https://github.com/elastic/elasticsearch/pull/66022#discussion_r557686188)).